otherlibs/win32unix/isatty.c

```diff
@@ -0,0 +1,22 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*                 David Allsopp, OCaml Labs, Cambridge.                  */
+/*                                                                        */
+/*   Copyright 2017 MetaStack Solutions Ltd.                              */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+#include <caml/mlvalues.h>
+#include "unixsupport.h"
+
+CAMLprim value unix_isatty(value fd)
+{
+  DWORD lpMode;
+  return (Val_bool(GetConsoleMode(Handle_val(fd), &lpMode)));
```

Comment:
This relies on the fact that GetConsoleMode fails (i.e. returns 0) iff the handle is neither for an input or output buffer of a console.  Are you confident that this always works?  I cannot find on MSDN a description of error cases.   (I could imagine e.g. that the function doesn't fail on some older versions of Windows even on non-console handles.)

Reply:
It's not explicit, I agree, but even for MSDN it would be very strange for a function to be documented as requiring a console handle and not to return `ERROR_INVALID_HANDLE` if it receives a handle to something else.

Reply:
In these useful cuddly open-source Microsoft times, it turns out we can often check these things: https://github.com/dotnet/corefx/blob/master/src/System.Console/src/System/ConsolePal.Windows.cs#L164

Reply:
That said, it's interesting that .NET first of all checks that the handle represents a character device.

Reply:
This was introduced in https://github.com/dotnet/corefx/commit/8cd8e7820a4c3fa60e3efaca2c87575f25c2c717.  Should we search advice from the author of this commit?

Reply:
You can't go wrong by copying the locig of the .NET code...

Reply:
Actually, I think we can go wrong - I'm going to risk asserting that the .NET code is actually incorrect. It was added in https://github.com/dotnet/corefx/commit/8cd8e7820a4c3fa60e3efaca2c87575f25c2c717#diff-6f40ebab17283a57f687629b80f8beb7 and it treats the output of [GetFileType](https://msdn.microsoft.com/en-us/library/windows/desktop/aa364960.aspx) as if it were a bitmask, which it is not. So the code they have will unnecessarily call GetConsoleMode for pipes as well.

I have pushed a commit which includes what I believe to a correct version of the test on GetFileType...